samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java

```diff
@@ -106,7 +110,10 @@ public void run() {
           callback.failure(new SamzaException(msg));
         }
       };
-      ScheduledFuture scheduledFuture = timer.schedule(timerTask, timeout, TimeUnit.MILLISECONDS);
+
+      final ScheduledFuture scheduledFuture = isDraining && (envelope.isDrain() || envelope.isWatermark())
+          ? timer.schedule(timerTask, drainCallbackTimeout, TimeUnit.MILLISECONDS)
+          : timer.schedule(timerTask, timeout, TimeUnit.MILLISECONDS);
```

Comment:
Can we add unit tests to the addition to ensure we use the drainTimeout vs timeout appropriately?

Reply:
Added